design/proposals/variable_schema_validation.md

```diff
@@ -0,0 +1,293 @@
+  - [Affected Areas](#affected-areas)
+  - [Backwards Compatibility](#backwards-compatibility)
+  - [Performance](#performance)
+  - [Security](#security)
+  - [Documentation](#documentation)
+  - [Version Update](#version-update)
+  - [Future Work](#future-work)
+  - [Delivery Plan](#delivery-plan)
+    - [Minimal Functionality](#minimal-functionality)
+    - [Extended Functionality](#extended-functionality)
+  
+## Introduction
+
+Conjur variables are resources that allow a user or an application to define and store sensitive information.
+
+In many cases, such sensitive information is comprised of multiple pieces that are tightly coupled and together represent a complete object. A few examples:
+
+- In order to access a MySQL database, a user needs a username, a password and an address, URL or connection string.
+- In order to access a REST API exposed by a certain service, a user might need either a username and password, a private key and a public certificate, or a token. In addition, the user will need the server public certificate and its address.
+
+Currently, this is handled by defining multiple variables in Conjur. Each variable contains a single value, such as a username, a password, etc. To bundle these pieces together, we use a naming convention. For example:
+
+```yaml
+- !policy
+  id: my-policy
+  body:
+  - &vars
+    - !variable mysql-db-creds/password
+    - !variable mysql-db-creds/username
+    - !variable mysql-db-creds/address
+    - !variable mysql-db-creds/port
+
+  - !host my-app
+  
+  - !permit
+    role: !host my-app
+    privileges: [ read,execute ]
+    resource: *vars
+```
+
+## Current Limitations
+
+- Multiple variables cannot be updated together, as a single transaction. This might lead to a momentary inconsitency when for example, a username is updated, the password should probabaly be updated as well.
+- There is no enforcement on the content of a variable, which can lead to unintended invalid content.
+- There is no enforcement on the group of variables needed for a certain use case. For example, a database connection requires a username, password, connection URL or address and port. A user could accidentally forget to specify a certain needed variable.
+
+## Proposed Solution
+
+The proposed solution is to add the ability to use JSON schemas and link them to variables, for which we want validation of their content. This will allow a single variable to store multiple, related values. While it's already possible today, to store JSON structured information in a variable, this proposed feature adds two added values:
+
+- The JSON structure is validated and enforced. If the input is invalid, the user will get a proper error message and the variable update will be denied.
+- The JSON structure will not be encrypted completely, but only the sensitive information. This will allow future seach capabilities based on the non sensitive information.
+
+### User Experience
```

Comment:
I think you should elaborate here on more scenarios, because we have here a new entity with a new relation to variables. For example:

1. Deleting schema - it is possible to delete a schema ? if it already defined to some variables ? if you want to return an error which contains all the variables under the schema you should save this info in the DB.
2. Creating new schema - i suppose we will have a validation on the schema itself 
3. Updating variable - what is the UX? for example first i performed get on the variable to find the schema id, and then i fetch the schema rules, and base on those rules i create a new variable object to update

Reply:
I added a section that explains about it. As a start, we will support only updating of the entire JSON structure. In the future, the variable API will support granular updates as well.

Reply:
And @nessiLahav, regarding your questions, I added appropriate sections in the proposal to answer your questions. Thanks for raising them, they are really good points.

Reply:
To summarize - there will be a an update of specific parts of json possible, and deleting them also, right